### PR TITLE
feat: Added addMetadata to token plugin

### PR DESCRIPTION
**Pull request by stegaBOB**

### CODE REVIEW

File: packages/js/src/plugins/tokenModule/TokenClient.ts [hunk 1]
```diff
@@ -45,10 +45,31 @@ import {
...
+  AddMetadataOutput,
+} from './addMetadata';
+import { MintWithMetadata } from '@/plugins';
 
 export class TokenClient {
   constructor(protected readonly metaplex: Metaplex) {}
 
+  addMetadata(
+    input: AddMetadataInput
+  ): Task<AddMetadataOutput & { mintWithMetadata: MintWithMetadata | Mint }> {
```
Comment (lorisleiva) [inline packages/js/src/plugins/tokenModule/TokenClient.ts:60]: If you're adding metadata and you end up with a `Mint` model then something went wrong.

```suggestion
  ): Task<AddMetadataOutput & { mintWithMetadata: MintWithMetadata }> {
```

You can assert that `mintWithMetadata` is not a `Mint` model by using `assertMintWithMetadata(...)`.

File: packages/js/src/plugins/tokenModule/TokenClient.ts [hunk 2]
```diff
@@ -45,10 +45,31 @@ import {
...
+  AddMetadataInput,
+  addMetadataOperation,
+  AddMetadataOutput,
+} from './addMetadata';
+import { MintWithMetadata } from '@/plugins';
 
 export class TokenClient {
   constructor(protected readonly metaplex: Metaplex) {}
 
+  addMetadata(
```
Comment (lorisleiva) [inline packages/js/src/plugins/tokenModule/TokenClient.ts:58]: As discussed on Slack, would you mind moving that operation to the NFT module instead?

File: packages/js/src/plugins/tokenModule/addMetadata.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  typeof Key,
+  AddMetadataInput,
+  AddMetadataOutput
+>;
+
+export interface AddMetadataInput {
+  // Accounts.
+  mint: PublicKey;
+  payer?: Signer; // Defaults to mx.identity().
+  updateAuthority?: Signer; // Defaults to mx.identity().
```
Comment (lorisleiva) [inline packages/js/src/plugins/tokenModule/addMetadata.ts:37]: Here the `updateAuthority` does not need to be a Signer.

```suggestion
  updateAuthority?: PublicKey; // Defaults to mx.identity().publicKey.
```

File: packages/js/src/plugins/tokenModule/addMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+> & {
+  createMetadataInstructionKey?: string;
+};
+
+export type AddMetadataBuilderContext = Omit<AddMetadataOutput, 'response'>;
+
+export const addMetadataBuilder = async (
+  metaplex: Metaplex,
+  params: AddMetadataBuilderParams
+): Promise<TransactionBuilder<AddMetadataBuilderContext>> => {
```
Comment (lorisleiva) [inline packages/js/src/plugins/tokenModule/addMetadata.ts:93]: There's no need for this builder method to be async.

**Review by lorisleiva — CHANGES_REQUESTED** [review]
Thanks! I left some comments.

**Comment by stegaBOB** [comment]
Resolved all the comments!

**Review by lorisleiva — APPROVED** [review]

**Comment by lorisleiva** [comment]
Thanks!

**Pull request merged**